src/ast/passes/c_macro_expansion.cpp

```diff
@@ -0,0 +1,76 @@
+#include "ast/passes/c_macro_expansion.h"
+#include "ast/ast.h"
+#include "ast/context.h"
+#include "ast/visitor.h"
+#include "clang_parser.h"
+#include "driver.h"
+
+namespace bpftrace::ast {
+
+class CMacroExpander : public Visitor<CMacroExpander> {
+public:
+  CMacroExpander(ASTContext &ast, CDefinitions &cdefinitions)
+      : ast_(ast), cdefinitions_(cdefinitions) {};
+
+  using Visitor<CMacroExpander>::visit;
+  void visit(Expression &expr);
+
+private:
+  ASTContext &ast_;
+  CDefinitions &cdefinitions_;
+  std::unordered_set<std::string> active_;
+};
+
+void CMacroExpander::visit(Expression &expr)
+{
+  // N.B. We only support raw identifier macros. The way expansion works is
+  // that we see if an expression is a bare identifier, then attempt expansion
+  // recurisvely.
+  if (auto ident = expr.as<Identifier>()) {
+    if (cdefinitions_.macros.contains(ident->ident)) {
+      auto &value = cdefinitions_.macros[ident->ident];
+
+      // Check for recursion.
+      if (active_.contains(ident->ident)) {
+        ident->addError() << "Macro recursion limit reached: " << ident->ident
```

Comment:
There really isn't a "limit" right? It's just the detection of any recursion.

Reply:
Yes, exactly. I just kept the message the same. The "limit" part is just because the current code expands until it has expanded 10000 characters, and then decides that it is recursion :P

I can change the message and tests.

Reply:
I've clarified this. It's actually much cleaner now with the "expanded from" context that you fixed :D